### PR TITLE
use hiera data in module instead of params.pp

### PR DESCRIPTION
**Pull request by Dan33l**
#### Pull Request (PR) description
Commons defaults values defined in `init.pp` and OS family specific values in hiera data of module.

### CODE REVIEW

File: data/os/RedHat.yaml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+---
+keepalived::sysconf_dir: 'sysconfig'
+keepalived::sysconf_options: '-D'
+keepalived::service_hasstatus: true
+keepalived::service_hasrestart: true
+
```
Comment (bastelfreak) [inline data/os/RedHat.yaml:6]: please remove those empty lines

File: manifests/init.pp [hunk 2]
```diff
@@ -1,36 +1,36 @@
 # == Class keepalived
 #
 class keepalived (
-  String[1] $sysconf_dir     = $keepalived::params::sysconf_dir,
-  String    $sysconf_options = $keepalived::params::sysconf_options,
+  String[1] $sysconf_dir,
+  String    $sysconf_options,
+  Boolean                 $service_hasrestart,
```
Comment (dhoppe) [inline manifests/init.pp:6]: Please fix the indentation.

Comment (Dan33l) [reply manifests/init.pp]: Yes, done.

**Comment by saz** [comment]
@dhoppe @bastelfreak The merge of this PR broke `keepalived::global_defs` as the value for `$keepalived::config_dir` is missing:
```
Warning: /Stage[main]/Keepalived::Global_defs/Concat::Fragment[keepalived.conf_globaldefs]/Concat_fragment[keepalived.conf_globaldefs]: Target Concat_file with path or title '/keepalived.conf' or tag '_keepalived.conf not found
 in the catalog
```

Trying to fix it results in "Duplicate declaration". I'm lost a bit here :(

**Comment by Dan33l** [comment]
Hi @saz,

Can you give a peace of puppet code that produce the warning ?

**Comment by saz** [comment]
@Dan33l Sure:

```
  class { 'keepalived::global_defs':
    script_user                   => 'nobody',
    enable_script_security => true,
    router_id                      => 'PUB',
  }
  class { 'keepalived':
    service_restart => 'killall -HUP keepalived',
  }
```
There's no other way of setting `script_user` for example.
This is working fine until the merge of this PR.

**Comment by saz** [comment]
btw this example is (more or less) given in the README itself.

**Comment by Dan33l** [comment]
I proposed a PR that add acceptance tests on global_defs: https://github.com/voxpupuli/puppet-keepalived/pull/217

It looks to work as expected. During tests it use `notification_email` that is a parameter OS independent.
@saz Can you reproduce with `notification_email` settings like in tests, or only with `script_user` ?

**Comment by saz** [comment]
I'm not sure why those tests should be working.
Looking at https://github.com/voxpupuli/puppet-keepalived/blob/4bdd57aec141cf79eec198b2b0ad5e62bfbce38b/manifests/global_defs.pp#L80 the used variable is missing.
Maybe it's a matter of ordering within the catalog?

**Comment by Dan33l** [comment]
The variable is existing here : https://github.com/voxpupuli/puppet-keepalived/blob/master/manifests/init.pp#L9
Or i missed something ?

**Comment by saz** [comment]
It's set in init, yes. Before, the class `keepalived` has been inherited.
After this PR, it's not the case anymore.

If you're looking at the error output in https://github.com/voxpupuli/puppet-keepalived/pull/210#issuecomment-648938256 you can see, that puppet is looking for `Concat_file` with target `'/keepalived.conf'` instead of `/etc/keepalived/keepalived.conf`.

Looking at the line in global_defs.pp:
`target  => "${keepalived::config_dir}/keepalived.conf",`
you can see, that the value of `${keepalived::config_dir}` is missing.

I've further tested this issue and it seems to be a matter of ordering within the manifest.

Working:
```
  class { 'keepalived':
    service_restart => 'killall -HUP keepalived',
  }

  class { 'keepalived::global_defs':
    script_user                    => 'nobody',
    enable_script_security => true,
    router_id                       => 'something',
  }
```

Broken:
```
  class { 'keepalived::global_defs':
    script_user                    => 'nobody',
    enable_script_security => true,
    router_id                       => 'something',
  }

  class { 'keepalived':
    service_restart => 'killall -HUP keepalived',
  }
```

Also not working (note the `before` parameter of class `keepalived`)
```
  class { 'keepalived::global_defs':
    script_user            => lookup('profile::keepalived::script_user', String, 'first', 'nobody'),
    enable_script_security => true,
    router_id              => lookup('profile::keepalived::router_id', Optional[String], 'first', undef),
  }

  class { 'keepalived':
    service_restart => 'killall -HUP keepalived',
    before          => Class['keepalived::global_defs'],
  }
```
I don't know if there's a way to fix this. From my pov it's more than unexpected to work like that

**Comment by Dan33l** [comment]
Something like that, in class `keepalived::global_defs` ?
```
  if ! defined(Class['keepalived']) {
    fail 'class ::keepalived has not been evaluated'
  }
```

**Comment by saz** [comment]
This will - at least - make puppet fail and the error more visible.
Not really sure, if it's the most user-friendly way, but maybe adding a hint to the documentation will help on this?

**Pull request merged**